Arm user mode

tyrantlizard77 opened the pull request:
Unfortunately, we have not been testing user mode well on non-x86. It is not tested by default on the platforms that support it.

This PR contains bug fixes and workaround for various issues. Finally, it ensures that user mode is enabled when running test cases for all platforms that support user mode, even if not switched on in the defconfig.

Commit:
tests: enable user mode by default

Many boards do not enable user mode in their default configuration
even though they support it. However for purposes of our unit
testing we really do want it on all the time.

It's not enough to just flip on CONFIG_USERSPACE, the tests largely
assume that CONFIG_APPLICATION_MEMORY and CONFIG_DYNAMIC_OBJECTS
are enabled. Update CONFIG_TEST_USERSPACE to be on by default if
CONFIG_TEST is active, and hang CONFIG_DYNAMIC_OBJECTS on it too.

This problem was not previously detected as qemu_x86 has the
user mode options in its defconfig and it is currently the only
user mode compatible platform which runs in emulation.

Fixes #7692

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
test_build_sensors_a_m: disable app memory

We run into the limit of 32 object files on ARM when
CONFIG_APPLICATION_MEMORY is enabled.

Bug #7703 filed, meanwhile just disable it, it's not
needed for this test case.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: mem_protect: fix off-by-one

A test was trying to add the maximum number of partitions,
but when the domain was initialized there was already one
added which needed to be accounted for to avoid an
assertion failing.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
elf_helper: fix member offset calculation

On some arches like ARC, the member location tag is a list with
the offset and then the member size. We just need the offset.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: prepare threads after PRE_KERNEL*

prepare_multithreading() was done very early as it had a call
to initialize the interrupt subsystem. This was causing problems
with stack pointer randomization as any HW-based entropy drivers
had not been initialized.

Move the call to initialize the interrupt system out of
prepare_multithreading(), which now really does just prepare
the system to start threads. This is now done after the PRE_KERNEL
phases.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: fatal: fix several issues

An errant commit accidentally disabled all testing of
hardware-based stack protection. Restore it, and work
around a problem with how these kinds of exceptions are
reported on ARM until #7706 is fixed.

We need to globally disable user mode due to how the
select statements in Kconfig work, the stack sentinel
is incompatible with user mode.

Some build warnings when compiling as native_posix
fixed.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
bluetooth: disable user mode for one test

This is a workaround until #7704 is fixed, we are getting
an assembler error due to the r7 clobber in the syscall
invocation macros.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arm: syscalls: fix some register issues

Upon return from a syscall handlers, the r1, r2, and r3 registers
could contain random kernel data that should not be leaked to user
mode. Zero these out before returning from _arm_do_syscall().
Fixes #7753.

The invocation macros need a clobber if r1, r2, or r3 are not used
to carry syscall arguments. This is a partial fix for #7754 but
there appear to be other issues.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andyross commented on scripts/elf_helper.py:
```diff
@@ -138,7 +138,10 @@ class AggregateTypeMember:
+        if isinstance(member_offset, list):
+            self.member_offset = member_offset[0]
+        else:
+            self.member_offset = member_offset
```
Yikes.  Is this like the worst API ever?  "It's a list.  Unless there is only one element, then I just give it to you".  Who does that?

andyross commented on tests/Kconfig:
```diff
@@ -36,9 +36,11 @@ config COVERAGE
 	select USERSPACE
 	select APPLICATION_MEMORY
-	default n
+	select DYNAMIC_OBJECTS
```
These two lines run afoul of me deep and persistent ignorance of the differences between "depends on" and "select" in kconfig.  I'm just going to nod, but secretly wish there was a way to express this without using both of them at once.

andyross commented on tests/kernel/fatal/src/main.c:
```diff
@@ -217,11 +210,15 @@ void test_fatal(void)
+#ifdef CONFIG_ARM
+	/* FIXME: See #7706 */
+	zassert_true(crash_reason == _NANO_ERR_STACK_CHK_FAIL ||
+		     crash_reason == _NANO_ERR_HW_EXCEPTION, NULL);
```
Is there really no way to distinguish STACK_CHK_FAIL from HW_EXCEPTION on ARM?  Or is this maybe because ARM uses HW_EXCEPTION as a catchall in a error handler that doesn't look quite finished?  I know xtensa does that, and they probably got it from somewhere...

andrewboie replied:
There is, but I don't have time to fix it, I filed a bug which I indicated in a comment. My interest here is to get past the breakage so we can enable all the tests.

andyross commented on tests/kernel/mem_protect/mem_protect/src/mem_domain.c:
```diff
@@ -313,8 +313,7 @@ void mem_domain_for_user_tc3(void *max_partitions, void *p2, void *p3)
 {
-
-	u8_t max_partitions = (u8_t)_arch_mem_domain_max_partitions_get();
+	u8_t max_partitions = (u8_t)_arch_mem_domain_max_partitions_get() - 1;
```
Probably worth a comment here explaining the -1.  Or is there a way to get the number of partitions allocated via the API?  You could use that .    The commit message makes sense, but someone looking at only the code isn't going to understand the need for that offset or the fact that one domain is already initialized.

andyross reviewed (approved):
One nitpick, a few notes & questions

MaureenHelm commented:
I ran this on both the frdm_k64f and frdm_kw41z boards and got the following results:
```
$ sanitycheck -T tests/kernel/ -p frdm_k64f --device-testing --device-serial /dev/ttyACM0
Cleaning output directory /home/maureen/zephyr/sanity-out
Building testcase defconfigs...
72 tests selected, 8595 tests discarded due to filters
total complete:   24/  72  33%  failed:    0

frdm_k64f                 tests/kernel/threads/lifecycle/thread_init/kernel.threads FAILED: timeout
	see: sanity-out/frdm_k64f/tests/kernel/threads/lifecycle/thread_init/kernel.threads/handler.log

total complete:   72/  72  100%  failed:    1
71 of 72 tests passed with 0 warnings in 1560 seconds
```
```
$ sanitycheck -T tests/kernel/ -p frdm_kw41z --device-testing --device-serial /dev/ttyACM0
Cleaning output directory /home/maureen/zephyr/sanity-out
Building testcase defconfigs...
63 tests selected, 8604 tests discarded due to filters
total complete:   24/  63  38%  failed:    0

frdm_kw41z                tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random FAILED: timeout
	see: sanity-out/frdm_kw41z/tests/kernel/mem_protect/stack_random/kernel.memory_protection.stack_random/handler.log

total complete:   63/  63  100%  failed:    1
62 of 63 tests passed with 0 warnings in 810 seconds
```

I'm not quite sure if the second failure is valid since this is a Cortex M0+ with no MPU.

Commit:
arm: userspace: fix initial user sp location

The initial user stack pointer gets rounded to STACK_ALIGN_SIZE,
which can be much finer granularity than MPU regions. With
certain stack size values passed to k_thread_create(), the stack
pointer can be set past the defined region for the user thread's
stack, causing an immediate MPU fault.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
